### ISSUE TITLE
zts_select with timeout on Windows exhausts semaphores

### ISSUE THREAD

**Issue by hach-que** [zerotier/libzt#25]
We're experiencing an issue with zts_select which results in semaphore exhaustion in lwip, which results in an assertion because the number of semaphores overflows.

Code to reproduce the issue can be found in ztserver on the "semaphore-broken" branch here: https://github.com/HiveMP/ztserver/blob/semaphore-broken/main.cpp#L240-L252

Once running the application, you can replicate it with:

```
curl -X PUT \
  http://localhost:8080/join \
  -H 'Cache-Control: no-cache' \
  -H 'Content-Type: application/json' \
  -H 'Postman-Token: 9350f4c5-4f53-c05d-4c3f-165a31f3f376' \
  -d '{
	"nwid": "8bd5124fd6206336",
	"path": "ztidentity2"
}'
```

Wait for ztserver to join the network, then run:

```
curl -X PUT \
  http://localhost:8080/forward \
  -H 'Cache-Control: no-cache' \
  -H 'Content-Type: application/json' \
  -H 'Postman-Token: 6080dc26-719e-0199-0fbf-4a64a9014c46' \
  -d '{
	"localport": 55000
}'
```

After a little while in the ztserver output, you'll see the following:

```
INFO [0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\ZT1Service.cpp:  225:        zts_start_service: binding to port=9995
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\ZT1Service.cpp:  454:                zts_start: waiting for node address assignment
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\ZT1Service.cpp:  512:          zts_get_node_id:
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\ZT1Service.cpp:  458:                zts_start: node=abba16d2f6
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\ZT1Service.cpp:  459:                zts_start: waiting for node to come online
INFO [0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\VirtualTap.cpp:   86:   VirtualTap::VirtualTap: set VirtualTap interface name to: libzt0-8bd5124fd6206336
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\lwIP.cpp:  218:         lwip_driver_init:
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\lwIP.cpp:  117:          tcpip_init_done: tcpip-thread
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\lwIP.cpp:  200:              main_thread: stack thread init complete
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\ZT1Service.cpp:  380:                 zts_join:
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\VirtualTap.cpp:  138:        VirtualTap::addIp: addr=172.25.1.66/16, nwid=8bd5124fd6206336
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\lwIP.cpp:  345:      lwip_init_interface: lwipInterfacesCount=0
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\lwIP.cpp:  347:      lwip_init_interface: netif=00EA0778
INFO [0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\lwIP.cpp:  373:      lwip_init_interface: initialized netif as [mac=36:c8:9a:c0:9d:e4, addr=172.25.1.66/16, nm=255.255.0.0/16]
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\VirtualTap.cpp:  138:        VirtualTap::addIp: addr=fc5d:f571:79ab:ba16:d2f6:0000:0000:0001/40, nwid=8bd5124fd6206336
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\lwIP.cpp:  345:      lwip_init_interface: lwipInterfacesCount=1
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\lwIP.cpp:  347:      lwip_init_interface: netif=00EA0828
INFO [0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\lwIP.cpp:  396:      lwip_init_interface: initialized netif as [mac=36:c8:9a:c0:9d:e4, addr=fc5d:f571:79ab:ba16:d2f6:0000:0000:0001/40]
TRANS[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\lwIP.cpp:  266:              lwip_eth_tx: len=   78 dst=33:33:ff:c0:9d:e4 [50df16d2f6 TX <-- abba16d2f6] proto=0x86dd IPv6
TRANS[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\lwIP.cpp:  266:              lwip_eth_tx: len=   78 dst=33:33:ff:00:00:01 [50dfd64f13 TX <-- abba16d2f6] proto=0x86dd IPv6
TRANS[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\lwIP.cpp:  266:              lwip_eth_tx: len=   62 dst=33:33:00:00:00:02 [5020d64f10 TX <-- abba16d2f6] proto=0x86dd IPv6
TRANS[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\lwIP.cpp:  266:              lwip_eth_tx: len=   70 dst=33:33:00:00:00:02 [5020d64f10 TX <-- abba16d2f6] proto=0x86dd IPv6
TRANS[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\lwIP.cpp:  266:              lwip_eth_tx: len=   70 dst=33:33:00:00:00:02 [5020d64f10 TX <-- abba16d2f6] proto=0x86dd IPv6
TRANS[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\lwIP.cpp:  266:              lwip_eth_tx: len=   70 dst=33:33:00:00:00:02 [5020d64f10 TX <-- abba16d2f6] proto=0x86dd IPv6
creating zerotier socket...
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\libzt.cpp:   63:               zts_socket: family=2, type=2, proto=0
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\libzt.cpp:   63:               zts_socket: family=23, type=2, proto=0
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\libzt.cpp:  110:                 zts_bind: fd=0
EXTRA[0]: C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\src\libzt.cpp:  110:                 zts_bind: fd=1
Assertion "sys_sem_free() closed more than created" failed at line 198 in C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\ext\lwip-contrib\ports\win32\sys_arch.c
Assertion "sys_sem_new() counter overflow" failed at line 176 in C:\RedpointJenkinsOSS\workspace\masteraf7ce1af@2\libzt\ext\lwip-contrib\ports\win32\sys_arch.c
```

and debugging in Visual Studio shows the breakpoint here:

![image](https://user-images.githubusercontent.com/504826/36102953-0683d404-1062-11e8-9d75-e38b9ee8dd37.png)

**Comment by joseph-henry** [comment]
Interesting project.

Try adding `-DLWIP_DEBUG=1` to the compile flags to get more insight on what might be happening with LWIP.

I've tried to replicate this but I don't yet see it on my test machines (macOS, CentOS and Windows 10).

How long does it take from the last `zts_bind()` statement before it overflows the semaphore counter? How many calls do you estimate it makes before failure?

Here's some dummy code I used to test for this issue, can you try the below code?:


```




I tried building your project but I can't seem to get it to properly produce a `ztserver.vcxproj`. 

Can you replicate the issue with a shorter snippet like the above? I don't have some of the dependencies on my system that your project requires.

**Comment by hach-que** [comment]
It does seem to vary in how long it takes. Sometimes it's around a minute, other times much much longer. But always the same error and always from zts_select.

I'll follow up trying the other diagnostics later today.

**Comment by joseph-henry** [comment]
Sounds good. I'll keep an eye out on my end.

**Comment by hach-que** [comment]
So I've managed to replicate the issue in ztserver on a second Windows 10 computer. You can get it to replicate in the ztserver code much faster if you set the timeout to something like this:

```
timeout.tv_sec = 0;
timeout.tv_usec = 100;
```

I tested with the following variants of your original test case, and got the following outcomes. You can view the test case variants here: https://github.com/HiveMP/ztserver/blob/semaphore-broken/zt_repl.cpp

- `_st` - Single-threaded
- `_mt` - Multi-threaded
- `_ipv4` - UDP IPv4 socket created and zts_select is used on it
- `_ipv6` - UDP IPv6 socket created and zts_select is used on it
- `_v4fd1` - UDP IPv4 socket created after the UDP IPv6 socket, resulting in UDP IPv4 having FD 1 instead of FD 0
- `_nfd0s` - The first socket (FD 0) is created, but we don't call zts_select on it (whether it's IPv4 or IPv6 depends on presence of `_v4fd1`)
- `_nfd1s` - The second socket (FD 1) is created, but we don't call zts_select on it (whether it's IPv4 or IPv6 depends on presence of `_v4fd1`)

All variants have a lowered timeout to make replication faster, and no `zts_accept` call because unlike the original code they're all UDP.

For multi-threaded variants, the `zts_bind` and `zts_listen` calls occur on the main thread, and the `zts_select` and `zts_recvfrom` occur on the child thread(s).

| Variant | Outcome |
| - | - |
| Original | Could not replicate |
| st_ipv4 | Could not replicate |
| st_ipv6 | Could not replicate |
| st_ipv4_ipv6 | Could not replicate |
| st_ipv4_ipv6_v4fd1 | Could not replicate |
| mt_ipv4 | Could not replicate |
| mt_ipv6 | Could not replicate |
| mt_ipv4_ipv6 | **Replicates** |
| mt_ipv4_ipv6_nfd0s | Could not replicate |
| mt_ipv4_ipv6_nfd1s | Could not replicate |
| mt_ipv4_ipv6_nfd0s_nfd1s | Could not replicate |
| mt_ipv4_ipv6_v4fd1 | **Replicates** |
| mt_ipv4_ipv6_v4fd1_nfd0s | Could not replicate |
| mt_ipv4_ipv6_v4fd1_nfd1s | Could not replicate |
| mt_ipv4_ipv6_v4fd1_nfd0s_nfd1s | Could not replicate |

Based on these outcomes, I think it's reasonable to make the assessment that this only **occurs when two different threads are both trying to call zts_select at the same time**.

Throughout testing I saw the following exception occur inside sys_thread_function, but I'm pretty sure it's unrelated to the issue at hand. It happened in most of the variants that I tested, and restarting the app seemed to work. I *think* that zts_select might be trying to free the timeout argument or something like that?

![image](https://user-images.githubusercontent.com/504826/36132730-f97af470-10cc-11e8-90c2-2d86a32f5e62.png)

**Comment by hach-que** [comment]
I've just noticed that the `fd_set` structure in the Windows headers, and the `fd_set` structure used by lwIP is different and incompatible.

I'm working on importing the lwIP headers into my project to make sure I'm passing lwIP's idea of fd_set into libzt instead of the Windows style structure.

**Comment by hach-que** [comment]
I've filed #26 for the differences in Winsock/lwIP structures and libzt's incorrect behaviour in directly passing the former into code that expects the latter.

**Comment by joseph-henry** [comment]
This is an old issue but one that I have deliberately not closed because I hadn't actually resolved it. However, there have been a few changes to the project since the opening of this ticket which might resolve this issue and I'm curious as to whether it still exists:

 - With the latest version I've provided better integration with the underlying network stack's functionality and exposed more of its native structures to the user via `ZeroTierSockets.h`. Previously I provided a sort of half-baked compatibility layer which wasn't very reliable. Now, if the user application sticks to using `zts_*` structures and `ZTS_*` constants defined in the C API header there's a lower chance of anything getting interpreted incorrectly.

- I've updated the lwIP submodule to 2.1.2 STABLE

If this remains a bug I am willing to re-approach it but it may be a bug/oversight in the lwIP code.

Let me know what help if any you'd like on this. And thanks again for your time and effort on this!
